feat: filter stores by name

kalleep opened the pull request:
<!-- Thanks for opening a PR! Here are some quick tips:
If this is your first time contributing, [read our Contributing Guidelines](https://github.com/openfga/.github/blob/main/CONTRIBUTING.md) to learn how to create an acceptable PR for this repo.
By submitting a PR to this repository, you agree to the terms within the [OpenFGA Code of Conduct](https://github.com/openfga/.github/blob/main/CODE_OF_CONDUCT.md)

If your PR is under active development, please submit it as a "draft". Once it's ready, open it up for review.
-->

<!-- Provide a brief summary of the changes -->

## Description
Add the ability to filter stores by name. This is useful when you want to have predictable identifiers for stores. While store names are not unique in openfga it can still be enforced in other ways. 

Right now the only option to find a store by name is to list all of them and do the filtering yourself. So having the ability to filter out stores by name is really useful.

To test this out locally you would need to check out the branch in this pr https://github.com/openfga/api/pull/211 and update go mod to point to you local repo 
```
replace github.com/openfga/api/proto => /path/to/proto
``` 

## References
Changes required in proto definition https://github.com/openfga/api/pull/211
Part of https://github.com/openfga/rfcs/pull/27

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

If you haven't done so yet, we would appreciate it if you could star the [OpenFGA repository](https://github.com/openfga/openfga). :)

jpadilla reviewed (changes requested):
@kalleep thanks! do you think we'll need/want/benefit from partial indexes that cover this query pattern?

jpadilla commented on pkg/server/commands/list_stores.go:
```diff
@@ -52,6 +52,7 @@ func (q *ListStoresQuery) Execute(ctx context.Context, req *openfgav1.ListStores
 
 	opts := storage.ListStoresOptions{
 		IDs:        storeIDs,
+		Name:       req.GetName(),
```
@kalleep can we temporarily point `github.com/openfga/api/proto` in this PR to https://github.com/openfga/api/pull/211  so that tests can actually run?

kalleep replied:
Done

rhamzeh commented on pkg/storage/test/stores.go:
```diff
@@ -83,6 +83,18 @@ func StoreTest(t *testing.T, datastore storage.OpenFGADatastore) {
 		require.Empty(t, ct)
 	})
 
+	t.Run("list_stores_succeeds_with_name_filter_match", func(t *testing.T) {
```
Just to add to this, may I ask we add tests for:
* multiple stores matching the name
* multiple stores matching the name crossing the pagination threshold to make sure we're paginating correctly when filtering by name?

kalleep replied:
I added test to cover these cases

kalleep commented on pkg/storage/storage.go:
```diff
@@ -87,6 +87,7 @@ type ReadAuthorizationModelsOptions struct {
 type ListStoresOptions struct {
 	// IDs is a list of store IDs to filter the results.
 	IDs        []string
+	Name       string
```
Yes I noticed that you had the pattern for that in some other places. I can make that split if you want

Looking at e.g. [ReadChanges](https://github.com/openfga/openfga/blob/8d794a97d56cc6778b6db2721a87c57703482444/pkg/storage/storage.go#L323) you would have leave pagination in `ListStoreOptions` and move `IDs` and `Name` to `ListStoresFilter` right?

Looking at your [releases](https://github.com/openfga/openfga/releases) these kind of changes are handled as a breaking change, but I get just adding `Name` of its own would be a breaking change.

jpadilla replied:
Nah let's not worry about this for now.

jpadilla reviewed (changes requested):
@kalleep looking good! Seems like linting is failing. We'll also want to add to this to the CHANGELOG.

There's some additional feedback on https://github.com/openfga/api/pull/211. Once we get that addressed and merged, feel free to mark this one as ready for review so we can get it merged in too!

kalleep commented:
@jpadilla I fixed the lint issues and added a new entry to the changelog.

Once https://github.com/openfga/api/pull/211 is merged I will remove https://github.com/openfga/openfga/pull/2103/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R141 and update to the latest version of the api

miparnisari reviewed (commented):
Could you update this test: https://github.com/openfga/openfga/blob/913adc90492fc54a2ae625dd52935b54ecb2620c/pkg/server/commands/list_stores_test.go#L32-L55

to ensure that the `name` parameter gets received by `mockDatastore`?

kalleep commented:
I can see that https://github.com/openfga/api/pull/211 is approved. When that one is merged I will update this pr to use the correct version of github.com/openfga/api/proto and mark it as ready
